### PR TITLE
catalog: Prune storage in the background

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, the startup process was blocked on reading in the entire
`mz_storage_usage_by_shard` table and pruning old entries. This commit
moves that pruning into a background task so startup is no longer
blocked.

Works towards resolving #28722

### Motivation
This PR adds a known-desirable feature.

### Tips for reviewer
* The diff is much smaller if viewed with whitespace hidden.
* Please look closely at the controller changes. I tried to do something clean with Futures and after fighting async for a couple of hours I gave up and did something messy.
* Review each commit separately.
* Don't let the LoC scare you, the majority is catalog migration plumbing.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/adapter/src/coord.rs [hunk 1]
```diff
@@ -2941,6 +2952,43 @@ impl Coordinator {
...
+            // TODO(jkosh44) Consider doing this periodically instead of only on startup.
+            if let Some(retention_period) = storage_usage_retention_period {
+                let id = self.catalog().resolve_builtin_table(&MZ_STORAGE_USAGE_BY_SHARD);
+                let read_ts = self.get_local_read_ts().await;
+                let current_contents_fut = self
+                    .controller
+                    .storage
+                    .snapshot_latest_async(id)
+                    .await
+                    .unwrap_or_terminate("cannot fail to fetch snapshot");
```
Comment (jkosh44) [inline src/adapter/src/coord.rs:2964]: I'm seeing a panic here for an unknown ID, maybe I should be using `self.controller.storage_collections` instead.

File: src/adapter/src/coord.rs [hunk 2]
```diff
@@ -3319,47 +3320,49 @@ impl Coordinator {
...
-            let collection_timestamp: u128 = collection_timestamp
-                .try_into()
-                .expect("all collections happen after Jan 1 1970");
-            if collection_timestamp < cutoff_ts {
-                debug!("pruning storage event {row:?}");
-                let builtin_update = BuiltinTableUpdate { id, row, diff: -1 };
-                expired.push(builtin_update);
+        let current_contents_fut = self.controller.storage.snapshot(id, read_ts);
+        let internal_cmd_tx = self.internal_cmd_tx.clone();
+        spawn(|| "storage usage prune", async move {
```
Comment (ParkMyCar) [inline src/adapter/src/coord.rs:3333]: Just want to double check that this only gets called once? I think ideally we'd prevent multiple instances of this task running at the same time?

**Review by ParkMyCar — APPROVED** [review]

File: src/adapter/src/coord.rs (hunk 2 above)
Comment (jkosh44) [reply src/adapter/src/coord.rs:3333]: Yes, once during startup. I'm hoping that the comment above this makes that clear.

File: src/adapter/src/coord/message_handler.rs [hunk 3]
```diff
@@ -321,6 +324,11 @@ impl Coordinator {
         }
     }
 
+    #[mz_ore::instrument(level = "debug")]
+    async fn storage_usage_prune(&mut self, expired: Vec<BuiltinTableUpdate>) {
+        self.builtin_table_update().execute(expired).await.await
```
Comment (aljoscha) [inline src/adapter/src/coord/message_handler.rs:329]: The dreaded double `.await.await` ... 🙈

File: src/storage-controller/src/lib.rs [hunk 4]
```diff
@@ -29,19 +29,14 @@ use itertools::Itertools;
 use mz_build_info::BuildInfo;
 use mz_cluster_client::client::ClusterReplicaLocation;
 use mz_cluster_client::ReplicaId;
-
 use mz_ore::collections::CollectionExt;
-use mz_persist_client::stats::{SnapshotPartsStats, SnapshotStats};
```
Comment (aljoscha) [inline src/storage-controller/src/lib.rs:34]: Random imports re-ordering, but fine for me... 😅

**Review by aljoscha — APPROVED** [review]

File: src/adapter/src/catalog/open.rs [hunk 5]
```diff
@@ -971,6 +948,7 @@ fn add_new_remove_old_builtin_items_migration(
         match system_object_mappings.remove(&desc) {
             Some(system_object_mapping) => {
                 if system_object_mapping.unique_identifier.fingerprint != fingerprint {
+                    assert_ne!(*MZ_STORAGE_USAGE_BY_SHARD_DESCRIPTION, system_object_mapping.description, "mz_storage_usage_by_shard cannot be migrated or else the table will be truncated");
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/adapter/src/catalog/open.rs]: Done.

File: src/storage-controller/src/lib.rs (hunk 4 above)
Comment (jkosh44) [reply src/storage-controller/src/lib.rs:34]: I originally did have changes in this file, but they were eventually pulled into a different PR. These imports were all out of order and my IDE will automatically order them if I change the file.

File: src/adapter/src/coord/message_handler.rs (hunk 3 above)
Comment (jkosh44) [reply src/adapter/src/coord/message_handler.rs:329]: I guess we should probably do the second await in a separate task so we don't block the main thread.

**Pull request merged**